crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,530 @@
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use crate::operation_text::hash;
+use common_lang_types::{
+    ArtifactFileName, ArtifactHash, ArtifactPathAndContent, FileContent, FileSystemOperation,
+    SelectableName, ServerObjectEntityName,
+};
+use isograph_config::PersistedDocumentsHashAlgorithm;
+
+#[derive(Debug, Clone, Default)]
+pub struct FileSystemState {
+    root_files: HashMap<ArtifactFileName, (FileContent, ArtifactHash)>,
+    nested_files: HashMap<
+        ServerObjectEntityName,
+        HashMap<SelectableName, HashMap<ArtifactFileName, (FileContent, ArtifactHash)>>,
+    >,
+}
+
+impl FileSystemState {
+    pub fn insert(&mut self, path_and_content: &ArtifactPathAndContent) {
```

Comment:
this should take ownership of `ArtifactPathAndContent`, so we don't have to clone.